File: src/Kriptografi.java
public class Kriptografi {
    public static void main(String[] args) throws Exception {
        System.out.println("Hello, World!");
        Orang amry = new Orang(2,19999);
        Orang reza = new Orang(2,19999);
        amry.mengirimPesan(reza, 1010);
    }
}

class Orang{
    private double privateKey;
    public double publicKey;
    private double id_number1;
    private double id_number2;

    public Orang(double id_number1, double id_number2) {
        this.privateKey = generatePrivateKey(id_number1,id_number2);
        this.publicKey = generatePublicKey(id_number1,id_number2);
        this.id_number1 = id_number1;
        this.id_number2 = id_number2;
    }

    public double gcd(double a,double b){
        return b==0 ? a : gcd(b, a%b); 
    }
    public double generatePrivateKey(double id_number1, double id_number2){
        double phi = (id_number1-1)*(id_number2-1);
        double e = 2;
        while (e < phi)
        {
            if (gcd(e, phi)==1)
                break;
            else
                e++;
        }
        int k = 2; 
        double d = (1 + (k*phi))/e;
        return d;
    }

    public double generatePublicKey(double id_number1, double id_number2){
        double e = 10;
        double phi = (id_number1-1)*(id_number2-1);
        while (e < phi)
        {
            if (gcd(e, phi)==1)
                break;
            else
                e++;
        }
        return e;
    }
    public void mengirimPesan(Orang penerima, double pesan){
        double enkripted = Math.pow(pesan, penerima.publicKey);
        enkripted %= penerima.id_number1*penerima.id_number2;
        System.out.println(enkripted);
        double dekripsi = Math.pow(enkripted, penerima.privateKey);
        dekripsi %= penerima.id_number1*penerima.id_number2;
        System.out.println(dekripsi);
    }
}

File: bin/kriptografi.py
class Orang:
    def __init__(self, id_1, id_2):
        self.private_key = self.generate_keys(id_1, id_2)[0]
        self.public_key = self.generate_keys(id_1, id_2)[1]
        self.n = id_1* id_2

    def generate_keys(self, p, q):
        n = p * q
        Pn = (p-1)*(q-1)
        key = []
        for i in range(2, Pn):
            gcd = self.euclid(Pn, i)
            if gcd == 1:
                key.append(i)
        e = int(313)
        r, d = self.exteuclid(Pn, e)
        return [e,d]

    def mengirim_pesan(self, penerima, pesan):
        pesan_terenkripsi = (pesan**penerima.public_key) % penerima.n
        print("pesan yang dikirim (ciphertext) :", pesan_terenkripsi)
        penerima.menerima_pesan(pesan_terenkripsi)

    def menerima_pesan(self, pesan_terenkripsi):
        pesan_terdekripsi = (pesan_terenkripsi**self.private_key) % self.n
        print("pesan terdekripsi :",pesan_terdekripsi)

    def euclid(self, m, n):
        if n == 0:
            return m
        else:
            r = m % n
            return self.euclid(n, r)

    def exteuclid(self, a, b):
        r1 = a
        r2 = b
        s1 = int(1)
        s2 = int(0)
        t1 = int(0)
        t2 = int(1)
        while r2 > 0:
            q = r1//r2
            r = r1-q * r2
            r1 = r2
            r2 = r
            s = s1-q * s2
            s1 = s2
            s2 = s
            t = t1-q * t2
            t1 = t2
            t2 = t
        if t1 < 0:
            t1 = t1 % a
        return (r1, t1)

class Pesan:
    konten = []
    def __init__(self, konten):
        for i in konten:
            self.konten.append(i)

amry = Orang(3, 357611)
reza = Orang(2, 349337)

pesan = 10110
print('pesan asli (plain text) : ', pesan)
print('\namry -> reza')
amry.mengirim_pesan(reza, pesan)
print('\nreza -> amry')
reza.mengirim_pesan(amry, pesan)